uniffi_bindgen/src/backend/types.rs

```diff
@@ -65,46 +55,46 @@ pub trait CodeType {
     /// This is the newer way of handling these methods and replaces the lower, write, lift, and
     /// read CodeType methods.  Currently only used by Kotlin, but the plan is to move other
     /// backends to using this.
-    fn ffi_converter_name(&self, oracle: &dyn CodeOracle) -> String {
-        oracle.class_name(&format!("FfiConverter{}", self.canonical_name(oracle)))
+    fn ffi_converter_name(&self) -> String {
+        format!("FfiConverter{}", self.canonical_name())
     }
 
     /// An expression for lowering a value into something we can pass over the FFI.
-    fn lower(&self, oracle: &dyn CodeOracle) -> String {
-        format!("{}.lower", self.ffi_converter_name(oracle))
+    fn lower(&self) -> String {
+        format!("{}.lower", self.ffi_converter_name())
     }
 
     /// An expression for writing a value into a byte buffer.
-    fn write(&self, oracle: &dyn CodeOracle) -> String {
-        format!("{}.write", self.ffi_converter_name(oracle))
+    fn write(&self) -> String {
+        format!("{}.write", self.ffi_converter_name())
     }
 
     /// An expression for lifting a value from something we received over the FFI.
-    fn lift(&self, oracle: &dyn CodeOracle) -> String {
-        format!("{}.lift", self.ffi_converter_name(oracle))
+    fn lift(&self) -> String {
+        format!("{}.lift", self.ffi_converter_name())
     }
 
     /// An expression for reading a value from a byte buffer.
-    fn read(&self, oracle: &dyn CodeOracle) -> String {
-        format!("{}.read", self.ffi_converter_name(oracle))
+    fn read(&self) -> String {
+        format!("{}.read", self.ffi_converter_name())
     }
 
     /// A list of imports that are needed if this type is in use.
     /// Classes are imported exactly once.
-    fn imports(&self, _oracle: &dyn CodeOracle) -> Option<Vec<String>> {
+    fn imports(&self) -> Option<Vec<String>> {
         None
     }
 
     /// Function to run at startup
-    fn initialization_fn(&self, _oracle: &dyn CodeOracle) -> Option<String> {
+    fn initialization_fn(&self) -> Option<String> {
         None
     }
 }
 
 /// This trait is used to implement `CodeType` for `Type` and type-like structs (`Record`, `Enum`, `Field`,
 /// etc).  We forward all method calls to a `Box<dyn CodeType>`, which we get by calling
 /// `CodeOracle.find()`.
-pub trait CodeTypeDispatch {
+pub trait CodeTypeDispatch: Debug {
```

Comment:
I would love to replace this trait with something like:

```
pub trait AsType {
   fn as_type(&self) -> Type;
}
```

The having filters that looked like:
```
pub fn type_name(t: &impl AsType) -> Result<String, askama::Error> {
    Ok(oracle.find(t.as_type()).type_label())
}
```

`AsType` is more general than `CodeTypeDispatch` and overall the code seems simpler to me.  Maybe this should be a separate PR though.

Reply:
That's actually a great idea and I agree it's much cleaner - I put up #1588, let me know if that's not what you had in mind.